README
Acme::CPANAuthors::MBTI

Minimal release atm.

Send bug reports ( or preferably patches ) if you want inclusion.

Will generate extra .pm files on demand as soon as we have members (
Trivial Effort )

Recommended quiz for people who have no idea and just want to play
along: <http://www.25quiz.com/>

Its not guaranteed to be accurate, but that doesn't matter.

Its a generalisation tool.

Just because the generalisation holds that averages of demographic
X[0..50] is "Y", that does not mean you can take X[1] and assert Y.

However, knowing X[1] is between X[0] and X[50] you can conjecture their
odds at being relative to the demographic average Y.

Thus, one can conjecture what one *might* do in a slightly better manner
than enumerating all possible behaviours observed in humans.

Blah blah blah more ranting here.

Creds to "ETHER" <https://metacpan.org/author/ETHER> for the original
entry I stole all the things from and refactored them to death,
Acme::CPANAuthors::Nonhuman

Subsequently, my source file is a lot smaller and needs about 3
variables modifed to make a new type.

So just change the files in "/authorlists"

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Acme-CPANAuthors-MBTI

0.001003 2017-03-07T16:25:05Z dbcd7fe

0.001002 2017-03-07T16:21:30Z 3994f29
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot
 - Fix build failure due to the same

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - develop: +5 ↑4 -2 (recommends: +1, suggests: ↑2)
 - test: ↓1

0.001001 2015-01-18T07:51:34Z 65b7481
 [Bugfix]
 - Have CPANAuthors that supports empty namespaces in ::CPANAuthors

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - configure: -1 (recommends: ↓1)
 - develop: +3 -1 (recommends: ↓1, suggests: ↑1)
 - runtime: +1
 - test: -1 (recommends: ↓2)

0.001000 2014-12-26T11:48:13Z 70a79bf
 - First version.


inc/expand_author_list.pm
use 5.014;    # /r modifier
use strict;
use warnings;

package expand_author_list;

# ABSTRACT: Expand a DATA section into a CPAN Authors list

# AUTHORITY

use HTTP::Tiny;
use Parse::CPAN::Whois;
use Acme::CPANAuthors::Factory;
use JSON::MaybeXS 1.001000;
use Path::Tiny qw(path);
use HTML::Entities;
use Math::Random::MT;
use List::MoreUtils qw( distinct );

sub tempdir { return state $tempdir = Path::Tiny->tempdir() }
sub http    { return state $http    = HTTP::Tiny->new() }
sub json    { return state $json    = JSON::MaybeXS->new( utf8 => 0 ) }

sub rng {
  return state $rng = Math::Random::MT->new( map { unpack 'L', $_ } q[Kent], q[ is ], q[a du], q[ll b], q[oy  ] );
}

sub dolog { *STDERR->print("[32m$_[0][0m\n") }

sub mirror_whois {
  my ($filename) = @_;
  my $out = tempdir()->child($filename);
  dolog("Fetching $filename to $out");
  my $response = http()->mirror( 'http://www.cpan.org/authors/' . $filename, $out );
  die "failed to fetch $filename: $response->{status} $response->{reason}\n"
    if not $response->{success} and $response->{status} ne '304';
  return "$out";
}

sub get_gravatar_url {
  my ($authorid) = @_;
  dolog("Resolving Gravatar for $authorid");
  my $request = http()->get( 'http://api.metacpan.org/v0/author/' . $authorid ) // {};
  my $content = $request->{content}                                             // {};
  my $json    = json()->decode($content);
  my $url     = $json->{gravatar_url};
  $url =~ s/s=\K130/80/g;
  return $url;
}

sub extract_data {
  state $result_cache = {};
  my ( $category, $source_file ) = @_;
  return $result_cache->{$category} if exists $result_cache->{$category};

  my $author_data = Parse::CPAN::Whois->new( mirror_whois('00whois.xml') );

  my $author_hash = {};
  my @authors;

  for my $id ( path($source_file)->lines_raw( { chomp => 1 } ) ) {
    dolog("$category / $id");
    my $name = $author_data->author($id)->name;
    $author_hash->{$id} = $name;
    push @authors,
      {
      id     => $id,
      name   => $name,
      avatar => get_gravatar_url($id),
      };
  }
  my $author_db = Acme::CPANAuthors::Factory->create( $category . '_temp' => $author_hash );
  for my $author (@authors) {
    $author->{dists} = $author_db->distributions( $author->{id} ) // 0;
  }

  return ( $result_cache->{$category} = [ sort { $b->{dists} <=> $a->{dists} || $a->{id} cmp $b->{id} } @authors ] );
}

sub authors_to_code {
  my (%config) = ref $_[0] ? %{ $_[0] } : @_;

  my $plugin         = $config{plugin};
  my $plugin_name    = $config{plugin_name} // ref $config{plugin};
  my $plugin_version = $config{plugin_version} // $plugin_name->VERSION;
  my $category       = $config{category};

  my $authors     = '';
  my $avatar_urls = '';

  for my $author ( @{ $config{data} } ) {
    $authors     .= "    $author->{id} => '$author->{name}',\n";
    $avatar_urls .= "    $author->{id} => '$author->{avatar}',\n";
  }

  my @display_authors = map { $config{data}->[ rng->irand() % scalar @{ $config{data} } ] } 1;

  return <<"EOF";
# Code inserted by inc/expand_author_list#authors_to_code
# by $plugin_name $plugin_version
## no critic (ValuesAndExpressions::RestrictLongStrings)
my \%authors  = (
$authors);

my \%avatar_urls = (
$avatar_urls);

## use critic

=method authors

  my \$scalar_ref = Acme::CPANAuthors::${category}\->authors;
  my \%hash       = Acme::CPANAuthors::${category}\->authors;

=cut

sub authors { return ( wantarray ? \%authors : \\\%authors ) }

=method category

  my \$scalar = Acme::CPANAuthors::${category}\->category;

=cut

sub category { return '$category' }

=method avatar_url

  my \$url = Acme::CPANAuthors::${category}\->avatar_url('$display_authors[0]->{id}');

=cut

sub avatar_url {
  my ( \$id ) = \@_;
  return \$avatar_urls{\$id};
}

# end generated code
EOF

}

sub authors_to_avatars {
  my (%config) = ref $_[0] ? %{ $_[0] } : @_;

  my $plugin         = $config{plugin};
  my $plugin_name    = $config{plugin_name} // ref $config{plugin};
  my $plugin_version = $config{plugin_version} // $plugin_name->VERSION;

  my @lines;
  for my $author ( @{ $config{data} } ) {
    my $name = encode_entities( $author->{name} );
    my $title = "$author->{id} ($name), $author->{dists} distribution" . ( $author->{dists} != 1 ? 's' : '' );
    push @lines,
        qq{<a href="http://metacpan.org/author/$author->{id}">}
      . q{<span>}
      . q{<img style="margin: 0 5px 5px 0;" width="80" height="80" }
      . qq{src="$author->{avatar}" alt="$author->{id}" title="$title" />}
      . q{</span>} . q{</a>};
  }
  my $content = join( "<!--\n-->", @lines );

  return <<"EOF";
<div style="text-align:center;padding:0px!important;overflow-y:hidden;
margin-left: auto; margin-right: auto; max-width: 430px">
<!-- Data inserted by inc/expand_author_list#authors_to_avatars
 by $plugin_name $plugin_version -->
$content
</div>
EOF

}

sub mbti_type {
  my ($type) = @_;
  return <<"EOF";
  L<< C<$type>|https://en.wikipedia.org/wiki/$type >>
EOF
}

sub mbti_description_text {
  my ($type) = @_;
  my $lctype = lc($type);
  return <<"EOF";
For more details see L<< C<Acme::CPANAuthors::MBTI>|Acme::CPANAuthors::MBTI >>.

=over 4

=item * L<< C<$type> on personalitypage.com|http://personalitypage.com/$type.html >>

=item * L<< C<$type> on typelogic.com|http://typelogic.com/$lctype.html >>

=item * L<< C<$type> on Wikipedia|https://en.wikipedia.org/wiki/$type >>

=back

EOF
}

sub mbti_description {
  my (%config) = ref $_[0] ? %{ $_[0] } : @_;

  my $html        = authors_to_avatars( \%config );
  my $description = mbti_description_text( $config{type} );
  my $link        = mbti_type( $config{type} );

  return <<"EOF";
This class provides a hash of PAUSE ID's and names of authors
who have identified themselves as $link

=begin html

$html

=end html

$description

EOF

}

sub generate_synopsis {
  my (%config) = ref $_[0] ? %{ $_[0] } : @_;

  my $plugin         = $config{plugin};
  my $plugin_name    = $config{plugin_name} // ref $config{plugin};
  my $plugin_version = $config{plugin_version} // $plugin_name->VERSION;
  my @items          = @{ $config{'data'} };
  my @authors        = map { $items[ rng->irand() % scalar @items ] } 1 .. 3;

  return ( <<"EOF" =~ s/^(\S)/    $1/msgr );
use Acme::CPANAuthors;
use Acme::CPANAuthors::$config{category};
# Or just use Acme::CPANAuthors::MBTI

my \$authors  = Acme::CPANAuthors->new('$config{category}');
my \$number   = \$authors->count;
my \@ids      = \$authors->id;
my \@distros  = \$authors->distributions('$authors[0]->{id}');
my \$url      = \$authors->avatar_url('$authors[1]->{id}');
my \$kwalitee = \$authors->kwalitee('$authors[2]->{id}');

my \%authorshash    = Acme::CPANAuthors::$config{category}\->authors;
my \$authorshashref = Acme::CPANAuthors::$config{category}\->authors;
my \$category       = Acme::CPANAuthors::$config{category}\->category;

EOF
}

1;


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

#my @stopwords = (qw());
#for my $wordlist (@stopwords) {
#  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
#}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

#$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
$bundle->remove_policy('CodeLayout::RequireUseUTF8');
$bundle->remove_policy('Tics::ProhibitLongLines');
$bundle->remove_policy('Documentation::PodSpelling'); # yolo
#$bundle->remove_policy('Documentation::RequirePodLinksIncludeText');

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



lib/Acme/CPANAuthors/MBTI.pm
use 5.006;    # our
use strict;
use warnings;

package Acme::CPANAuthors::MBTI;

our $VERSION = '0.001003';

# ABSTRACT: CPAN Authors by MBTI Type

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Acme::CPANAuthors::MBTI::INTP;



































1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Acme::CPANAuthors::MBTI - CPAN Authors by MBTI Type

=head1 VERSION

version 0.001003

=head1 DESCRIPTION

Minimal release atm.

Send bug reports ( or preferably patches ) if you want inclusion.

Will generate extra .pm files on demand as soon as we have members ( Trivial Effort )

Recommended quiz for people who have no idea and just want to play along: L<http://www.25quiz.com/>

Its not guaranteed to be accurate, but that doesn't matter.

Its a generalisation tool.

Just because the generalisation holds that averages of demographic X[0..50] is "Y",
that does not mean you can take X[1] and assert Y.

However, knowing X[1] is between X[0] and X[50] you can conjecture their odds
at being relative to the demographic average Y.

Thus, one can conjecture what one I<might> do in a slightly better manner than enumerating
all possible behaviours observed in humans.

Blah blah blah more ranting here.

Creds to L<< C<ETHER>|https://metacpan.org/author/ETHER >> for the original entry I stole
all the things from and refactored them to death, L<Acme::CPANAuthors::Nonhuman>

Subsequently, my source file is a lot smaller and needs about 3 variables modifed to make a new type.

So just change the files in C</authorlists>

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Acme/CPANAuthors/MBTI/INTP.pm
use 5.008;    # utf8
use strict;
use warnings;
use utf8;

package Acme::CPANAuthors::MBTI::INTP;

# ABSTRACT: The TiNe CPAN Authors

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

our $VERSION = '0.001003';
#<<<
# Code inserted by inc/expand_author_list#authors_to_code
# by Dist::Zilla::Plugin::MungeFile::WithDataSection 0.009
## no critic (ValuesAndExpressions::RestrictLongStrings)
my %authors  = (
    KENTNL => 'Kent Fredric',
);

my %avatar_urls = (
    KENTNL => 'https://secure.gravatar.com/avatar/4b96677ae9ecbe76b8c33ead244f7407?s=80&d=http%3A%2F%2Fwww.gravatar.com%2Favatar%2F4b96677ae9ecbe76b8c33ead244f7407%3Fs%3D130%26d%3Didenticon',
);

## use critic








sub authors { return ( wantarray ? %authors : \%authors ) }







sub category { return 'MBTI::INTP' }







sub avatar_url {
  my ( $id ) = @_;
  return $avatar_urls{$id};
}

# end generated code

#>>>

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Acme::CPANAuthors::MBTI::INTP - The TiNe CPAN Authors

=head1 VERSION

version 0.001003

=head1 SYNOPSIS

    use Acme::CPANAuthors;
    use Acme::CPANAuthors::MBTI::INTP;
    # Or just use Acme::CPANAuthors::MBTI

    my $authors  = Acme::CPANAuthors->new('MBTI::INTP');
    my $number   = $authors->count;
    my @ids      = $authors->id;
    my @distros  = $authors->distributions('KENTNL');
    my $url      = $authors->avatar_url('KENTNL');
    my $kwalitee = $authors->kwalitee('KENTNL');

    my %authorshash    = Acme::CPANAuthors::MBTI::INTP->authors;
    my $authorshashref = Acme::CPANAuthors::MBTI::INTP->authors;
    my $category       = Acme::CPANAuthors::MBTI::INTP->category;

=head1 DESCRIPTION

This class provides a hash of PAUSE ID's and names of authors
who have identified themselves as   L<< C<INTP>|https://en.wikipedia.org/wiki/INTP >>

=head1 METHODS

=head2 authors

  my $scalar_ref = Acme::CPANAuthors::MBTI::INTP->authors;
  my %hash       = Acme::CPANAuthors::MBTI::INTP->authors;

=head2 category

  my $scalar = Acme::CPANAuthors::MBTI::INTP->category;

=head2 avatar_url

  my $url = Acme::CPANAuthors::MBTI::INTP->avatar_url('KENTNL');

=for html <div style="text-align:center;padding:0px!important;overflow-y:hidden;
margin-left: auto; margin-right: auto; max-width: 430px">
<!-- Data inserted by inc/expand_author_list#authors_to_avatars
 by Dist::Zilla::Plugin::MungeFile::WithDataSection 0.009 -->
<a href="http://metacpan.org/author/KENTNL"><span><img style="margin: 0 5px 5px 0;" width="80" height="80" src="https://secure.gravatar.com/avatar/4b96677ae9ecbe76b8c33ead244f7407?s=80&d=http%3A%2F%2Fwww.gravatar.com%2Favatar%2F4b96677ae9ecbe76b8c33ead244f7407%3Fs%3D130%26d%3Didenticon" alt="KENTNL" title="KENTNL (Kent Fredric), 178 distributions" /></span></a>
</div>

For more details see L<< C<Acme::CPANAuthors::MBTI>|Acme::CPANAuthors::MBTI >>.

=over 4

=item * L<< C<INTP> on personalitypage.com|http://personalitypage.com/INTP.html >>

=item * L<< C<INTP> on typelogic.com|http://typelogic.com/intp.html >>

=item * L<< C<INTP> on Wikipedia|https://en.wikipedia.org/wiki/INTP >>

=back

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "CPAN Authors by MBTI Type",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Acme-CPANAuthors-MBTI",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "Acme::CPANAuthors::MBTI",
  "PREREQ_PM" => {
    "Acme::CPANAuthors" => "0.24",
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.001003",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Acme::CPANAuthors" => "0.24",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
